File: ios/MarketerCloudPersonalizationRNModule.swift
import ExpoModulesCore
import Evergage

public class MarketerCloudPersonalizationRNModule: Module {
    
  private var activeCampaigns: [String: EVGCampaign] = [:]
  private var listeners: Set<String> = []
    
  private func campaignHandler (campaign: EVGCampaign) {
      
      NSLog("MarketerCloudPersonalizationRNModule: campaignHandler: \(campaign)")
      
      if listeners.contains(campaign.target) {
          activeCampaigns[campaign.campaignId] = campaign
          
          let payload: [String: Any] = [
            "target": campaign.target,
            "campaignId": campaign.campaignId,
            "campaignName": campaign.campaignName,
            "experienceId": campaign.experienceId,
            "experienceName": campaign.experienceName,
            "messageId": campaign.messageId,
            "isControlGroup": campaign.isControlGroup,
            "data": campaign.data
          ]
          
          self.sendEvent("mcp_data_campaign", payload)
      }
  }
    
  public func definition() -> ModuleDefinition {
      Name("MarketerCloudPersonalizationRN")

      Events("mcp_data_campaign")
                  
      AsyncFunction("stopListeningForCampaignTarget") { (campaignTarget: String) in
          let evergage = Evergage.sharedInstance()
          if let context = evergage.globalContext {
              context.setCampaignHandler(nil, forTarget: campaignTarget)
          }
          
          if self.listeners.contains(campaignTarget) {
              self.listeners.remove(campaignTarget)
          }
      }
      
      AsyncFunction("listenForCampaignTarget") { (campaignTarget: String) in
        let evergage = Evergage.sharedInstance()
        weak var weakSelf = self
        if let context = evergage.globalContext {
          context.setCampaignHandler(weakSelf?.campaignHandler, forTarget: campaignTarget)
        }

        self.listeners.insert(campaignTarget)
      }
      
      AsyncFunction("getAccountId") {
        let evergage = Evergage.sharedInstance()
        return evergage.accountId
      }
    
      AsyncFunction("setAccountId") { (value: String) in
        let evergage = Evergage.sharedInstance()
        evergage.accountId = value
      }
                    
      AsyncFunction("getAnonymousId") {
        let evergage = Evergage.sharedInstance()
        return evergage.anonymousId
      }
      
      AsyncFunction("getUserId") {
        let evergage = Evergage.sharedInstance()
        return evergage.userId
      }
                    
      Function("setUserId") { (value: String) in
        let evergage = Evergage.sharedInstance()
        evergage.userId = value
      }
      
      AsyncFunction("viewItem") { (id: String) in
        let evergage = Evergage.sharedInstance()
        if let context = evergage.globalContext {
            context.viewItem(EVGProduct(id: id))
        }
      }
      
      AsyncFunction("viewCategory") { (id: String) in
        let evergage = Evergage.sharedInstance()
        if let context = evergage.globalContext {
            context.viewCategory(EVGCategory(id: id))
        }
      }
      
      AsyncFunction("trackAction") { (action: String) in
        let evergage = Evergage.sharedInstance()
        if let context = evergage.globalContext {
          context.trackAction(action)
        }
      }
      
      AsyncFunction("addToCart") { (productId: String, quantity: Int) in
        let evergage = Evergage.sharedInstance()
        if let context = evergage.globalContext {
            context.add(toCart: EVGLineItem(item: EVGProduct(id: productId), quantity: NSNumber(value: quantity)))
        }
      }
      
      AsyncFunction("trackClickthrough") { (campaignId: String) in
        let evergage = Evergage.sharedInstance()
        if let context = evergage.globalContext {
            if let c = activeCampaigns[campaignId] {
                context.trackClickthrough(c)
            }
        }
      }
      
      AsyncFunction("trackDismissal") { (campaignId: String) in
          let evergage = Evergage.sharedInstance()
          if let context = evergage.globalContext {
              if let c = activeCampaigns[campaignId] {
                  context.trackDismissal(c)
              }
          }
      }
      
      AsyncFunction("trackImpression") { (campaignId: String) in
          let evergage = Evergage.sharedInstance()
          if let context = evergage.globalContext {
              if let c = activeCampaigns[campaignId] {
                  context.trackImpression(c)
              }
          }
      }
  }
}

File: android/src/main/java/expo/modules/mobilemarketercloudpersonalization/MarketerCloudPersonalizationRNModule.kt
package expo.modules.mobilemarketercloudpersonalization

import com.evergage.android.Campaign
import com.evergage.android.Evergage
import com.evergage.android.promote.Category
import com.evergage.android.promote.LineItem
import com.evergage.android.promote.Product
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.concurrent.ConcurrentHashMap
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken

class MarketerCloudPersonalizationRNModule : Module() {

  private var listeners: MutableSet<String> = mutableSetOf();
  private var activeCampaigns: MutableMap<String, Campaign> = ConcurrentHashMap<String, Campaign>();

  override fun definition() = ModuleDefinition {
    Name("MarketerCloudPersonalizationRN")
    Events("mcp_data_campaign")

    AsyncFunction("stopListeningForCampaignTarget") { campaignTarget: String ->
      val evergage = Evergage.getInstance()
      val context = evergage.globalContext
      context?.setCampaignHandler(null, campaignTarget)

      if(listeners.contains(campaignTarget)) {
        listeners.remove(campaignTarget)
      }
    }

    AsyncFunction("listenForCampaignTarget") { campaignTarget: String ->
      val evergage = Evergage.getInstance()
      val context = evergage.globalContext

      val handler = { campaign: Campaign ->
          activeCampaigns[campaign.campaignId] = campaign;

          val gson = Gson()
          val type = object : TypeToken<Map<String, Any>>() {}.type
          val data: Map<String, Any> = gson.fromJson(campaign.data.toString(), type)

          sendEvent("mcp_data_campaign", mapOf(
              "target" to campaignTarget,
              "campaignId" to campaign.campaignId,
              "campaignName" to campaign.campaignName,
              "experienceId" to campaign.experienceId,
              "experienceName" to campaign.experienceName,
              "messageId" to campaign.messageId,
              "isControlGroup" to campaign.isControlGroup,
              "data" to data
          ))
      }

      context?.setCampaignHandler(handler, campaignTarget)
      listeners.add(campaignTarget)
    }

    AsyncFunction("getAccountId") {
        return@AsyncFunction Evergage.getInstance().accountId
    }

    AsyncFunction("setAccountId") { value: String ->
      Evergage.getInstance().accountId = value
    }

    AsyncFunction("getAnonymousId") {
        return@AsyncFunction Evergage.getInstance().anonymousId
    }

    AsyncFunction("getUserId") {
        return@AsyncFunction Evergage.getInstance().userId
    }

    Function("setUserId") { value: String ->
      Evergage.getInstance().userId = value
    }

    AsyncFunction("viewItem") { id: String ->
     Evergage.getInstance().globalContext?.viewItem(Product(id));
    }

    AsyncFunction("viewCategory") { id: String ->
      Evergage.getInstance().globalContext?.viewCategory(Category(id))
    }

    AsyncFunction("trackAction") { action: String ->
     Evergage.getInstance().globalContext?.trackAction(action)
    }

    AsyncFunction("addToCart") { productId: String, quantity: Int ->
      Evergage.getInstance().globalContext?.addToCart(LineItem(Product(productId), quantity))
    }

    AsyncFunction("trackClickthrough") { campaignId: String ->
        activeCampaigns[campaignId]?.let {
          Evergage.getInstance().globalContext?.trackClickthrough(it)
        }
    }

    AsyncFunction("trackDismissal") { campaignId: String ->
      activeCampaigns[campaignId]?.let {
        Evergage.getInstance().globalContext?.trackDismissal(it);
      }
    }

    AsyncFunction("trackImpression") { campaignId: String ->
      activeCampaigns[campaignId]?.let {
        Evergage.getInstance().globalContext?.trackImpression(it);
      }
    }

  }
}
